### 1. Objective
   ⇉ Goal of this project is to **predict the sales** of a Retail outlet   based on the historical data provided for sales.


### 2. Column Description

   ⇉ **Item Identifier**: A code provided for the item of sale <br>
   ⇉ **Item Weight**: Weight of item <br>
   ⇉ **Item Fat Content**: A categorical column of how much fat is present in the item : ‘Low Fat’, ‘Regular’, ‘low fat’, ‘LF’, ‘reg’ <br>
   ⇉ **Item Visibility**: Numeric value for how visible the item is  <br> 
   ⇉ **Item Type**: What category does the item belong to: ‘Dairy’, ‘Soft Drinks’, ‘Meat’, ‘Fruits and Vegetables’, ‘Household’, ‘Baking Goods’, ‘Snack Foods’, ‘Frozen Foods’, ‘Breakfast’, ’Health and Hygiene’, ‘Hard Drinks’, ‘Canned’, ‘Breads’, ‘Starchy Foods’, ‘Others’, ‘Seafood’. <br>
   ⇉ **Item MRP**: The MRP price of item <br>
   ⇉ **Outlet Identifier**: Which outlet was the item sold. This will be categorical column <br>
   ⇉ **Outlet Establishment Year**: Which year was the outlet established <br>
   ⇉ **Outlet Size**: A categorical column to explain size of outlet: ‘Medium’, ‘High’, ‘Small’.  <br> 
   ⇉ **Outlet Location Type**: A categorical column to describe the location of the outlet: ‘Tier 1’, ‘Tier 2’, ‘Tier 3’  <br>
   ⇉ **Outlet Type** : Categorical column for type of outlet: ‘Supermarket Type1’, ‘Supermarket Type2’, ‘Supermarket Type3’, ‘Grocery Store’  <br>
   ⇉ **Item Outlet Sales**: The amount of sales for an item.  <br> 
   ⇉ **Source**: Whether the data is from train or test.  <br>
   


   
### 3. Initial Exploratory Data Analysis

In [1]:
# Importing the packages and dataset

import numpy as np
import pandas as pd
import matplotlib as mlb
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
sns.set_style("whitegrid")

In [2]:
data_Train = pd.read_csv("Train_Retail.csv")
data_Test = pd.read_csv("Test_Retail.csv")
data_Train['Source'] = 'train'
data_Test['Source'] = 'test'

In [3]:
data_Test.shape

(5681, 12)

In [4]:
data_Train.shape

(8523, 13)

In [5]:
data_Train.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales Source  
0  Supermarket Type1          3735.1380  train  
1  Supermarket Type2           443.4228  train  
2  Supermarket Type1          2097.2700  train  
3      Grocery Store           732.3800  train  
4  Supermarket Type1           994.7052  train

In [6]:
data_Test.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  \
0           FDW58       20.750          Low Fat         0.007565  Snack Foods   
1           FDW14        8.300              reg         0.038428        Dairy   
2           NCN55       14.600          Low Fat         0.099575       Others   
3           FDQ58        7.315          Low Fat         0.015388  Snack Foods   
4           FDY38          NaN          Regular         0.118599        Dairy   

   Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
0  107.8622            OUT049                       1999      Medium   
1   87.3198            OUT017                       2007         NaN   
2  241.7538            OUT010                       1998         NaN   
3  155.0340            OUT017                       2007         NaN   
4  234.2300            OUT027                       1985      Medium   

  Outlet_Location_Type        Outlet_Type Source  
0               Tier 1  Supermarket Type1   test  
1               Tier 2  Supermarket Type1   test  
2               Tier 3      Grocery Store   test  
3               Tier 2  Supermarket Type1   test  
4               Tier 3  Supermarket Type3   test

In [7]:
data_Test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5681 entries, 0 to 5680
Data columns (total 12 columns):
Item_Identifier              5681 non-null object
Item_Weight                  4705 non-null float64
Item_Fat_Content             5681 non-null object
Item_Visibility              5681 non-null float64
Item_Type                    5681 non-null object
Item_MRP                     5681 non-null float64
Outlet_Identifier            5681 non-null object
Outlet_Establishment_Year    5681 non-null int64
Outlet_Size                  4075 non-null object
Outlet_Location_Type         5681 non-null object
Outlet_Type                  5681 non-null object
Source                       5681 non-null object
dtypes: float64(3), int64(1), object(8)
memory usage: 532.7+ KB


In [8]:
data_Train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 13 columns):
Item_Identifier              8523 non-null object
Item_Weight                  7060 non-null float64
Item_Fat_Content             8523 non-null object
Item_Visibility              8523 non-null float64
Item_Type                    8523 non-null object
Item_MRP                     8523 non-null float64
Outlet_Identifier            8523 non-null object
Outlet_Establishment_Year    8523 non-null int64
Outlet_Size                  6113 non-null object
Outlet_Location_Type         8523 non-null object
Outlet_Type                  8523 non-null object
Item_Outlet_Sales            8523 non-null float64
Source                       8523 non-null object
dtypes: float64(4), int64(1), object(8)
memory usage: 865.7+ KB


In [9]:
import pandas_profiling

In [10]:
#Report = pandas_profiling.ProfileReport(data_Train)

#Report.to_file("Retail_train_data.html")

#### Some Important Observations from the Training dataset:

Train Dataset have missing data values for the columns **"Item_Weight"** and **"Outlet_Size"** each of them having **1463** and **2410** missing values.

Item_Identifier has a high cardinality: **1559 distinct values**<br>
Item_Visibility has **526 / 6.2% zeros**<br>
Item_Weight has **17.2% missing values**<br>
Outlet_Size has **28.3% missing values**

In [11]:
#Report = pandas_profiling.ProfileReport(data_Test)

#Report.to_file("Retail_test_data.html")

#### Some Important Observations from the Testing dataset:

Test Dataset have missing data values for the columns **"Item_Weight"** and **"Outlet_Size"** each of them having **976** and **1606** missing values.

Item_Identifier has a high cardinality: **1543 distinct values**<br>
Item_Visibility has **353 / 6.2% zeros**<br>
Item_Weight has **17.2% missing values**<br>
Outlet_Size has **28.3% missing values**

In [12]:
data_Train.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  7060.000000      8523.000000  8523.000000                8523.000000   
mean     12.857645         0.066132   140.992782                1997.831867   
std       4.643456         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       8.773750         0.026989    93.826500                1987.000000   
50%      12.600000         0.053931   143.012800                1999.000000   
75%      16.850000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800

1. Column "Item_Weight" has values ranging from 4.56 to 21.35 with a mean value of 12.86 (adjusted to 2 decimal points). Count is showing as 7060 although we have 8523 number of rows in our training dataset. This phenomenon is justified by missing value count of 1463.
2. "Item_MRP" has values ranging from 31.29 to 266.89 with a mean value of 141. 
3. "Outlet_Establishment_Year" has values ranging from 1985 to 2009. From this we can see that, first store started is in 1985 and the most recent store started in 2009.

In [13]:
data_Train.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  7060.000000      8523.000000  8523.000000                8523.000000   
mean     12.857645         0.066132   140.992782                1997.831867   
std       4.643456         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       8.773750         0.026989    93.826500                1987.000000   
50%      12.600000         0.053931   143.012800                1999.000000   
75%      16.850000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800

### 4. Dealing with Missing values in the dataset

In [14]:
data_Train.mean()

Item_Weight                    12.857645
Item_Visibility                 0.066132
Item_MRP                      140.992782
Outlet_Establishment_Year    1997.831867
Item_Outlet_Sales            2181.288914
dtype: float64

In [15]:
data_Train.groupby("Item_Type")["Item_Weight"].mean()

Item_Type
Baking Goods             12.277108
Breads                   11.346936
Breakfast                12.768202
Canned                   12.305705
Dairy                    13.426069
Frozen Foods             12.867061
Fruits and Vegetables    13.224769
Hard Drinks              11.400328
Health and Hygiene       13.142314
Household                13.384736
Meat                     12.817344
Others                   13.853285
Seafood                  12.552843
Snack Foods              12.987880
Soft Drinks              11.847460
Starchy Foods            13.690731
Name: Item_Weight, dtype: float64

In [16]:
def fil_val(feat_var1):
    if pd.isna(feat_var1["Item_Weight"]):
        return data_Train[data_Train["Item_Type"] == feat_var1["Item_Type"]]["Item_Weight"].mean()
    else:
        return feat_var1["Item_Weight"]

In [17]:
data_Train['Item_Weight'].fillna(data_Train[["Item_Type", "Item_Weight"]].apply(fil_val, axis=1), inplace=True)
#data_Train[["Item_Type", "Item_Weight"]].apply(fil_val, axis=1)

In [18]:
data_Test['Item_Weight'].fillna(data_Train[["Item_Type", "Item_Weight"]].apply(fil_val, axis=1), inplace=True)

In [19]:
data_Train.head(10)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15      9.30000          Low Fat         0.016047   
1           DRC01      5.92000          Regular         0.019278   
2           FDN15     17.50000          Low Fat         0.016760   
3           FDX07     19.20000          Regular         0.000000   
4           NCD19      8.93000          Low Fat         0.000000   
5           FDP36     10.39500          Regular         0.000000   
6           FDO10     13.65000          Regular         0.012741   
7           FDP10     12.98788          Low Fat         0.127470   
8           FDH17     16.20000          Regular         0.016687   
9           FDU28     19.20000          Regular         0.094450   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   
5           Baking Goods   51.4008            OUT018   
6            Snack Foods   57.6588            OUT013   
7            Snack Foods  107.7622            OUT027   
8           Frozen Foods   96.9726            OUT045   
9           Frozen Foods  187.8214            OUT017   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   
5                       2009      Medium               Tier 3   
6                       1987        High               Tier 3   
7                       1985      Medium               Tier 3   
8                       2002         NaN               Tier 2   
9                       2007         NaN               Tier 2   

         Outlet_Type  Item_Outlet_Sales Source  
0  Supermarket Type1          3735.1380  train  
1  Supermarket Type2           443.4228  train  
2  Supermarket Type1          2097.2700  train  
3      Grocery Store           732.3800  train  
4  Supermarket Type1           994.7052  train  
5  Supermarket Type2           556.6088  train  
6  Supermarket Type1           343.5528  train  
7  Supermarket Type3          4022.7636  train  
8  Supermarket Type1          1076.5986  train  
9  Supermarket Type1          4710.5350  train

In [20]:
data_Test.isnull().sum()

Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  1606
Outlet_Location_Type            0
Outlet_Type                     0
Source                          0
dtype: int64

In [21]:
pd.crosstab(data_Train['Outlet_Size'], 'count')

col_0        count
Outlet_Size       
High           932
Medium        2793
Small         2388

In [22]:
#data_Train[data_Train["Outlet_Size"].isnull()]
data_Train[data_Train["Item_Outlet_Sales"] >= 2500]['Outlet_Location_Type'].value_counts()

Tier 3    1245
Tier 2    1073
Tier 1     712
Name: Outlet_Location_Type, dtype: int64

In [23]:
data_Train.loc[:,'Outlet_Size':'Outlet_Type']

Outlet_Size Outlet_Location_Type        Outlet_Type
0         Medium               Tier 1  Supermarket Type1
1         Medium               Tier 3  Supermarket Type2
2         Medium               Tier 1  Supermarket Type1
3            NaN               Tier 3      Grocery Store
4           High               Tier 3  Supermarket Type1
5         Medium               Tier 3  Supermarket Type2
6           High               Tier 3  Supermarket Type1
7         Medium               Tier 3  Supermarket Type3
8            NaN               Tier 2  Supermarket Type1
9            NaN               Tier 2  Supermarket Type1
10        Medium               Tier 1  Supermarket Type1
11         Small               Tier 1  Supermarket Type1
12        Medium               Tier 1  Supermarket Type1
13         Small               Tier 1  Supermarket Type1
14          High               Tier 3  Supermarket Type1
15         Small               Tier 1  Supermarket Type1
16        Medium               Tier 3  Supermarket Type2
17        Medium               Tier 1  Supermarket Type1
18        Medium               Tier 3  Supermarket Type3
19         Small               Tier 2  Supermarket Type1
20          High               Tier 3  Supermarket Type1
21        Medium               Tier 3  Supermarket Type3
22         Small               Tier 2  Supermarket Type1
23         Small               Tier 1      Grocery Store
24         Small               Tier 1  Supermarket Type1
25           NaN               Tier 2  Supermarket Type1
26         Small               Tier 2  Supermarket Type1
27          High               Tier 3  Supermarket Type1
28           NaN               Tier 3      Grocery Store
29         Small               Tier 1      Grocery Store
...          ...                  ...                ...
8493         NaN               Tier 2  Supermarket Type1
8494         NaN               Tier 2  Supermarket Type1
8495       Small               Tier 2  Supermarket Type1
8496         NaN               Tier 2  Supermarket Type1
8497       Small               Tier 2  Supermarket Type1
8498      Medium               Tier 3  Supermarket Type2
8499       Small               Tier 2  Supermarket Type1
8500         NaN               Tier 2  Supermarket Type1
8501         NaN               Tier 2  Supermarket Type1
8502         NaN               Tier 2  Supermarket Type1
8503       Small               Tier 2  Supermarket Type1
8504      Medium               Tier 3  Supermarket Type3
8505      Medium               Tier 3  Supermarket Type2
8506      Medium               Tier 3  Supermarket Type2
8507       Small               Tier 2  Supermarket Type1
8508         NaN               Tier 2  Supermarket Type1
8509         NaN               Tier 3      Grocery Store
8510       Small               Tier 2  Supermarket Type1
8511      Medium               Tier 3  Supermarket Type2
8512        High               Tier 3  Supermarket Type1
8513       Small               Tier 2  Supermarket Type1
8514         NaN               Tier 2  Supermarket Type1
8515      Medium               Tier 3  Supermarket Type2
8516      Medium               Tier 3  Supermarket Type2
8517       Small               Tier 1  Supermarket Type1
8518        High               Tier 3  Supermarket Type1
8519         NaN               Tier 2  Supermarket Type1
8520       Small               Tier 2  Supermarket Type1
8521      Medium               Tier 3  Supermarket Type2
8522       Small               Tier 1  Supermarket Type1

[8523 rows x 3 columns]

In [24]:
pd.crosstab(data_Train.Outlet_Location_Type, [data_Train.Outlet_Size, data_Train.Outlet_Type], margins=True)

Outlet_Size                       High            Medium                    \
Outlet_Type          Supermarket Type1 Supermarket Type1 Supermarket Type2   
Outlet_Location_Type                                                         
Tier 1                               0               930                 0   
Tier 2                               0                 0                 0   
Tier 3                             932                 0               928   
All                                932               930               928   

Outlet_Size                                    Small                     All  
Outlet_Type          Supermarket Type3 Grocery Store Supermarket Type1        
Outlet_Location_Type                                                          
Tier 1                               0           528               930  2388  
Tier 2                               0             0               930   930  
Tier 3                             935             0                 0  2795  
All                                935           528              1860  6113

In [25]:
data_Train["Outlet_Size"].fillna("High", inplace=True)
data_Test["Outlet_Size"].fillna("High", inplace=True)

In [26]:
data_Test.isnull().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Source                       0
dtype: int64

In [27]:
data_Train.isnull().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
Source                       0
dtype: int64

### Visualizations

In [28]:
# Lets Begin With Pairplot
# Visualizing the relation between the variables
sns.set()
sns.pairplot(data_Train)
#mlb.get_backend()

**Observation:** There is no significant correlation between the variables.

In [29]:
# Heatmap

corr = data_Train.corr()

sns.heatmap(corr, annot=True)

**Observations**: There is moderate correlation between **Item_Outlet_Sales** and **Item_MRP**.

In [30]:
sns.lmplot('Item_Outlet_Sales', 'Item_MRP', data=data_Train)

In [31]:
plt.figure(figsize=(15,6))
plt.xticks(rotation=90)
#plt.ylim(2200,2500)
sns.barplot(x='Item_Type', y = 'Item_Outlet_Sales', data = data_Train)

**Observation** : Most of the sales amount are happening for **Starchy Foods** or **Seafood**. 

In [32]:
plt.figure(figsize=(15,6))
plt.xticks(rotation=90)
sns.countplot(data_Train.Item_Type, hue=data_Train.Outlet_Location_Type)

**Observation**: Fruits and Vegetables are the most sold items in Tier 3 and Tier 1 cities whereas Snack Foods are the most sold items in Tier 2 cities. <br>

So finally we can conclude that Fruits and Vegetables are the most sold items in 3 location and Seafood is the least sold item in 3 locations.

In [33]:
data_Train.groupby(['Outlet_Location_Type'])['Item_Type'].value_counts()

Outlet_Location_Type  Item_Type            
Tier 1                Fruits and Vegetables    337
                      Snack Foods              334
                      Frozen Foods             250
                      Household                250
                      Dairy                    205
                      Baking Goods             185
                      Canned                   177
                      Health and Hygiene       138
                      Meat                     123
                      Soft Drinks              120
                      Breads                    67
                      Hard Drinks               58
                      Others                    56
                      Starchy Foods             36
                      Breakfast                 33
                      Seafood                   19
Tier 2                Snack Foods              401
                      Fruits and Vegetables    399
                      Household       

**Observation** : There is not much difference in purchase patterns in different city types.

In [34]:
sns.boxplot(data_Train.Outlet_Size, data_Train.Item_Outlet_Sales)

In [35]:
sns.boxenplot(data_Train.Outlet_Size, data_Train.Item_Outlet_Sales)

**Observation**: Maximum sales is incurred from medium size outlets. Outliers are present in the Item_Outlet_Sales across all the given Outlet_Size.<br>

Median outlet sales of Medium outlet size varies greatly from High and Small sized outlets and that can be confirmed from the table below.

In [36]:
Median_Sales = data_Train.groupby("Outlet_Size")["Item_Outlet_Sales"].median()
Median_Sales

Outlet_Size
High      1637.8680
Medium    2251.0698
Small     1544.6560
Name: Item_Outlet_Sales, dtype: float64

In [37]:
sns.barplot(x = 'Outlet_Size', y = 'Item_Outlet_Sales', data = data_Train)

**Observation** : Maximum sales is observed in Mid Sized Outlets.

In [38]:
plt.figure(figsize=(12,6))
sns.barplot(x='Outlet_Type', y = 'Item_Outlet_Sales', data = data_Train)

**Observation** :  Most of the sales are hapenning in Supermarket Type 3.

In [39]:
plt.figure(figsize=(12,6))
sns.boxplot(x='Outlet_Establishment_Year', y = 'Item_Outlet_Sales', data = data_Train)

**Observation:** Outliers are observed across all the years and the least sales is observed in the Outlet Establishment year 1998

In [40]:
plt.figure(figsize=(12,6))
sns.countplot(data_Train.Outlet_Establishment_Year)

**Observation**: In the dataset we have few transactions of outlets established in 1998. <br>
This implies that there is less footfalls in outlets established in 1998 and therefore the sales is believed the least. 

In [41]:
plt.figure(figsize=(12,6))
sns.barplot(x='Outlet_Location_Type', y = 'Item_Outlet_Sales', data = data_Train)

**Observation** : Maximum sales is observed in Tier 2 and 3 cities. Now we should inspect more for low sales in Tier 1 cities.

In [42]:
data_Train.Outlet_Location_Type.value_counts()

Tier 3    3350
Tier 2    2785
Tier 1    2388
Name: Outlet_Location_Type, dtype: int64

**Observation** : The number of outlets in Tier 1 cities is less than that of Tier 2 and 3 cities. There are more people living in Tier 1 cities and we can expect more Sales over there. But this is not observed in the above graph.<br>
<br>
**Let us find what factors are influencing the low sales in Tier 1 cities.**

In [43]:
data_Train.groupby('Outlet_Location_Type')['Item_MRP'].mean()

Outlet_Location_Type
Tier 1    140.870106
Tier 2    141.167196
Tier 3    140.935232
Name: Item_MRP, dtype: float64

**Observation**: We wanted to inspect whether Item_MRP has any impact in Low Sales in Tier 1 cities, but from the table above
we observe that there average MRP across 3 different Outlet_Location_Type is similar.

In [44]:
sns.boxplot(x='Outlet_Location_Type', y = 'Item_Outlet_Sales', hue='Outlet_Size', data = data_Train)

**Observation**: There are no big outlets in Tier 1 cities. <br>
    This gives us the reason why there is less sales in Tier 1 cities.

### 5. Feature Engineering

Convert **Outlet_Size** as it is nominal

In [45]:
data_Train["Outlet_Size"] = data_Train["Outlet_Size"].replace({'Small': 0, 'Medium': 1, 'High': 2})

In [46]:
data_Test["Outlet_Size"] = data_Test["Outlet_Size"].replace({'Small': 0, 'Medium': 1, 'High': 2})

In [47]:
data_Train.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year  Outlet_Size Outlet_Location_Type  \
0                       1999            1               Tier 1   
1                       2009            1               Tier 3   
2                       1999            1               Tier 1   
3                       1998            2               Tier 3   
4                       1987            2               Tier 3   

         Outlet_Type  Item_Outlet_Sales Source  
0  Supermarket Type1          3735.1380  train  
1  Supermarket Type2           443.4228  train  
2  Supermarket Type1          2097.2700  train  
3      Grocery Store           732.3800  train  
4  Supermarket Type1           994.7052  train

In [48]:
data_Test.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  \
0           FDW58       20.750          Low Fat         0.007565  Snack Foods   
1           FDW14        8.300              reg         0.038428        Dairy   
2           NCN55       14.600          Low Fat         0.099575       Others   
3           FDQ58        7.315          Low Fat         0.015388  Snack Foods   
4           FDY38        8.930          Regular         0.118599        Dairy   

   Item_MRP Outlet_Identifier  Outlet_Establishment_Year  Outlet_Size  \
0  107.8622            OUT049                       1999            1   
1   87.3198            OUT017                       2007            2   
2  241.7538            OUT010                       1998            2   
3  155.0340            OUT017                       2007            2   
4  234.2300            OUT027                       1985            1   

  Outlet_Location_Type        Outlet_Type Source  
0               Tier 1  Supermarket Type1   test  
1               Tier 2  Supermarket Type1   test  
2               Tier 3      Grocery Store   test  
3               Tier 2  Supermarket Type1   test  
4               Tier 3  Supermarket Type3   test

In [49]:
data_Train.Item_Identifier.tail()

8518    FDF22
8519    FDS36
8520    NCJ29
8521    FDN46
8522    DRG01
Name: Item_Identifier, dtype: object

#### Classify the items into broad categories - Creating a derived attribute

**Let's derive Item_Category based on the Item_Identifier.**

If we check the column Item_Identifier keenly we can observe that, Item_Identifier has a code prefixed in it based on the category which they belogns to(FDA15, DRC01, FDX07, NCD19...etc.). Broadly there are three categories of items each of which are identified by FD, DR and NC. Here I'm naming the derived column as Item_Category. 

In [50]:
data_Train["Item_Category"] = data_Train["Item_Identifier"].apply(lambda x: x[0:2])

In [51]:
data_Test["Item_Category"] = data_Test["Item_Identifier"].apply(lambda x: x[0:2])

In [52]:
data_Train.Item_Category.value_counts()

FD    6125
NC    1599
DR     799
Name: Item_Category, dtype: int64

Lets observe which columns require more data cleaning

In [53]:
data_Train.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year  Outlet_Size Outlet_Location_Type  \
0                       1999            1               Tier 1   
1                       2009            1               Tier 3   
2                       1999            1               Tier 1   
3                       1998            2               Tier 3   
4                       1987            2               Tier 3   

         Outlet_Type  Item_Outlet_Sales Source Item_Category  
0  Supermarket Type1          3735.1380  train            FD  
1  Supermarket Type2           443.4228  train            DR  
2  Supermarket Type1          2097.2700  train            FD  
3      Grocery Store           732.3800  train            FD  
4  Supermarket Type1           994.7052  train            NC

In [54]:
data_Train.Item_Fat_Content.value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

We can see that column Item_Fat_Content has two different values indicating each amount of Fat Content such as **Low Fat and LF** for items with low value of Fat content. Similarly with the **Regular and reg** for regular fat content items.  <br>
We are going to replace these multiple categories into 2 categories.

In [55]:
data_Train['Item_Fat_Content'] = data_Train['Item_Fat_Content'].replace({'LF':'Low Fat','reg':'Regular','low fat':'Low Fat'})
data_Train.reset_index(level=0, inplace=True)

#Check if they are replaced properly
data_Train.Item_Fat_Content.unique()

array(['Low Fat', 'Regular'], dtype=object)

In [56]:
data_Test['Item_Fat_Content'] = data_Test['Item_Fat_Content'].replace({'LF':'Low Fat','reg':'Regular','low fat':'Low Fat'})
data_Test.reset_index(level=0, inplace=True)

#Check if they are replaced properly
data_Test.Item_Fat_Content.unique()

array(['Low Fat', 'Regular'], dtype=object)

In [57]:
data_Train.head()

index Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0      0           FDA15         9.30          Low Fat         0.016047   
1      1           DRC01         5.92          Regular         0.019278   
2      2           FDN15        17.50          Low Fat         0.016760   
3      3           FDX07        19.20          Regular         0.000000   
4      4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year  Outlet_Size Outlet_Location_Type  \
0                       1999            1               Tier 1   
1                       2009            1               Tier 3   
2                       1999            1               Tier 1   
3                       1998            2               Tier 3   
4                       1987            2               Tier 3   

         Outlet_Type  Item_Outlet_Sales Source Item_Category  
0  Supermarket Type1          3735.1380  train            FD  
1  Supermarket Type2           443.4228  train            DR  
2  Supermarket Type1          2097.2700  train            FD  
3      Grocery Store           732.3800  train            FD  
4  Supermarket Type1           994.7052  train            NC

In [58]:
data_Test.head()

index Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0      0           FDW58       20.750          Low Fat         0.007565   
1      1           FDW14        8.300          Regular         0.038428   
2      2           NCN55       14.600          Low Fat         0.099575   
3      3           FDQ58        7.315          Low Fat         0.015388   
4      4           FDY38        8.930          Regular         0.118599   

     Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
0  Snack Foods  107.8622            OUT049                       1999   
1        Dairy   87.3198            OUT017                       2007   
2       Others  241.7538            OUT010                       1998   
3  Snack Foods  155.0340            OUT017                       2007   
4        Dairy  234.2300            OUT027                       1985   

   Outlet_Size Outlet_Location_Type        Outlet_Type Source Item_Category  
0            1               Tier 1  Supermarket Type1   test            FD  
1            2               Tier 2  Supermarket Type1   test            FD  
2            2               Tier 3      Grocery Store   test            NC  
3            2               Tier 2  Supermarket Type1   test            FD  
4            1               Tier 3  Supermarket Type3   test            FD

Now we have two types of fat contents for all the items, however we know from the previous steps that we have non-consumable items. Having fat content for those items doesn't make sense. Let's create a new category for them. 

In [59]:
data_Train.loc[data_Train['Item_Category']=='NC', 'Item_Fat_Content'] = "Non-Edible"

In [60]:
data_Test.loc[data_Test['Item_Category']=='NC', 'Item_Fat_Content'] = "Non-Edible"

In [61]:
data_Test.Item_Fat_Content.unique()

array(['Low Fat', 'Regular', 'Non-Edible'], dtype=object)

In [62]:
pd.crosstab(data_Train.Outlet_Size, 'count')

col_0        count
Outlet_Size       
0             2388
1             2793
2             3342

### 6. Handling Categorical Variable

In [63]:
# Lets drop the columns which have been feature engineered and have no effect on the model

data_Train.drop(['index', 'Item_Identifier', 'Outlet_Identifier', 'Source', 'Outlet_Establishment_Year'], axis=1, inplace=True)
data_Train.shape

(8523, 10)

In [64]:
data_Train.head()

Item_Weight Item_Fat_Content  Item_Visibility              Item_Type  \
0         9.30          Low Fat         0.016047                  Dairy   
1         5.92          Regular         0.019278            Soft Drinks   
2        17.50          Low Fat         0.016760                   Meat   
3        19.20          Regular         0.000000  Fruits and Vegetables   
4         8.93       Non-Edible         0.000000              Household   

   Item_MRP  Outlet_Size Outlet_Location_Type        Outlet_Type  \
0  249.8092            1               Tier 1  Supermarket Type1   
1   48.2692            1               Tier 3  Supermarket Type2   
2  141.6180            1               Tier 1  Supermarket Type1   
3  182.0950            2               Tier 3      Grocery Store   
4   53.8614            2               Tier 3  Supermarket Type1   

   Item_Outlet_Sales Item_Category  
0          3735.1380            FD  
1           443.4228            DR  
2          2097.2700            FD  
3           732.3800            FD  
4           994.7052            NC

In [65]:
# Lets drop the columns which have been feature engineered and have no effect on the model

data_Test.drop(['index', 'Item_Identifier', 'Outlet_Identifier', 'Source', 'Outlet_Establishment_Year'], axis=1, inplace=True)
data_Test.shape

(5681, 9)

In [66]:
data_Test.head()

Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  Item_MRP  \
0       20.750          Low Fat         0.007565  Snack Foods  107.8622   
1        8.300          Regular         0.038428        Dairy   87.3198   
2       14.600       Non-Edible         0.099575       Others  241.7538   
3        7.315          Low Fat         0.015388  Snack Foods  155.0340   
4        8.930          Regular         0.118599        Dairy  234.2300   

   Outlet_Size Outlet_Location_Type        Outlet_Type Item_Category  
0            1               Tier 1  Supermarket Type1            FD  
1            2               Tier 2  Supermarket Type1            FD  
2            2               Tier 3      Grocery Store            NC  
3            2               Tier 2  Supermarket Type1            FD  
4            1               Tier 3  Supermarket Type3            FD

In [67]:
# All the categorical columns

cat_feat_var = ['Item_Fat_Content','Outlet_Location_Type','Outlet_Size','Item_Category','Outlet_Type','Item_Type']

In [68]:
data_Train.columns

Index(['Item_Weight', 'Item_Fat_Content', 'Item_Visibility', 'Item_Type',
       'Item_MRP', 'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type',
       'Item_Outlet_Sales', 'Item_Category'],
      dtype='object')

In [69]:
categorical_df = data_Train[cat_feat_var]

In [70]:
categorical_df.head()

Item_Fat_Content Outlet_Location_Type  Outlet_Size Item_Category  \
0          Low Fat               Tier 1            1            FD   
1          Regular               Tier 3            1            DR   
2          Low Fat               Tier 1            1            FD   
3          Regular               Tier 3            2            FD   
4       Non-Edible               Tier 3            2            NC   

         Outlet_Type              Item_Type  
0  Supermarket Type1                  Dairy  
1  Supermarket Type2            Soft Drinks  
2  Supermarket Type1                   Meat  
3      Grocery Store  Fruits and Vegetables  
4  Supermarket Type1              Household

In [71]:
categorical_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 6 columns):
Item_Fat_Content        8523 non-null object
Outlet_Location_Type    8523 non-null object
Outlet_Size             8523 non-null int64
Item_Category           8523 non-null object
Outlet_Type             8523 non-null object
Item_Type               8523 non-null object
dtypes: int64(1), object(5)
memory usage: 399.6+ KB


In [72]:
categorical_df['Outlet_Size'] = categorical_df['Outlet_Size'].astype(str)

C:\Users\vishw\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [73]:
# Lets check the value counts of the Outlet_Size

pd.crosstab(categorical_df.Outlet_Size,1)

col_0           1
Outlet_Size      
0            2388
1            2793
2            3342

In [74]:
# DUMIFYING COLUMNS STARTS IN 3...2...1!!!

dummy_train_df = pd.get_dummies(categorical_df, drop_first=True) 

In [75]:
dummy_train_df.shape

(8523, 26)

In [76]:
dummy_train_df.columns

Index(['Item_Fat_Content_Non-Edible', 'Item_Fat_Content_Regular',
       'Outlet_Location_Type_Tier 2', 'Outlet_Location_Type_Tier 3',
       'Outlet_Size_1', 'Outlet_Size_2', 'Item_Category_FD',
       'Item_Category_NC', 'Outlet_Type_Supermarket Type1',
       'Outlet_Type_Supermarket Type2', 'Outlet_Type_Supermarket Type3',
       'Item_Type_Breads', 'Item_Type_Breakfast', 'Item_Type_Canned',
       'Item_Type_Dairy', 'Item_Type_Frozen Foods',
       'Item_Type_Fruits and Vegetables', 'Item_Type_Hard Drinks',
       'Item_Type_Health and Hygiene', 'Item_Type_Household', 'Item_Type_Meat',
       'Item_Type_Others', 'Item_Type_Seafood', 'Item_Type_Snack Foods',
       'Item_Type_Soft Drinks', 'Item_Type_Starchy Foods'],
      dtype='object')

In [77]:
# Concat data_Train and dummy_train_df

onehot_encoded_data = pd.concat([data_Train, dummy_train_df], axis= 1)

In [78]:
onehot_encoded_data.columns

Index(['Item_Weight', 'Item_Fat_Content', 'Item_Visibility', 'Item_Type',
       'Item_MRP', 'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type',
       'Item_Outlet_Sales', 'Item_Category', 'Item_Fat_Content_Non-Edible',
       'Item_Fat_Content_Regular', 'Outlet_Location_Type_Tier 2',
       'Outlet_Location_Type_Tier 3', 'Outlet_Size_1', 'Outlet_Size_2',
       'Item_Category_FD', 'Item_Category_NC', 'Outlet_Type_Supermarket Type1',
       'Outlet_Type_Supermarket Type2', 'Outlet_Type_Supermarket Type3',
       'Item_Type_Breads', 'Item_Type_Breakfast', 'Item_Type_Canned',
       'Item_Type_Dairy', 'Item_Type_Frozen Foods',
       'Item_Type_Fruits and Vegetables', 'Item_Type_Hard Drinks',
       'Item_Type_Health and Hygiene', 'Item_Type_Household', 'Item_Type_Meat',
       'Item_Type_Others', 'Item_Type_Seafood', 'Item_Type_Snack Foods',
       'Item_Type_Soft Drinks', 'Item_Type_Starchy Foods'],
      dtype='object')

In [79]:
onehot_encoded_data.shape

(8523, 36)

In [80]:
onehot_encoded_data.drop(data_Train[cat_feat_var], axis=1, inplace=True)
onehot_encoded_data.shape

(8523, 30)

In [81]:
onehot_encoded_data.shape

(8523, 30)

### 7. Creating Train and Validation Data

In this step we are going to create data models that are capable of predicting the Outlet_Sales. Inorder to create these models first we split the given Train dataset into Train and Validation datasets. Train dataset is the one which have the ItemOutlet_Sales values. We will train the model based on these values to validate the Validation dataset values.


In this first we are going to create a basic Linear Regression model.

In [82]:
# Lets split the training features and label

data_Train_Y = onehot_encoded_data["Item_Outlet_Sales"]
del onehot_encoded_data["Item_Outlet_Sales"]
data_Train_X = onehot_encoded_data

In [83]:
data_Train_X.head()

Item_Weight  Item_Visibility  Item_MRP  Item_Fat_Content_Non-Edible  \
0         9.30         0.016047  249.8092                            0   
1         5.92         0.019278   48.2692                            0   
2        17.50         0.016760  141.6180                            0   
3        19.20         0.000000  182.0950                            0   
4         8.93         0.000000   53.8614                            1   

   Item_Fat_Content_Regular  Outlet_Location_Type_Tier 2  \
0                         0                            0   
1                         1                            0   
2                         0                            0   
3                         1                            0   
4                         0                            0   

   Outlet_Location_Type_Tier 3  Outlet_Size_1  Outlet_Size_2  \
0                            0              1              0   
1                            1              1              0   
2                            0              1              0   
3                            1              0              1   
4                            1              0              1   

   Item_Category_FD  ...  Item_Type_Fruits and Vegetables  \
0                 1  ...                                0   
1                 0  ...                                0   
2                 1  ...                                0   
3                 1  ...                                1   
4                 0  ...                                0   

   Item_Type_Hard Drinks  Item_Type_Health and Hygiene  Item_Type_Household  \
0                      0                             0                    0   
1                      0                             0                    0   
2                      0                             0                    0   
3                      0                             0                    0   
4                      0                             0                    1   

   Item_Type_Meat  Item_Type_Others  Item_Type_Seafood  Item_Type_Snack Foods  \
0               0                 0                  0                      0   
1               0                 0                  0                      0   
2               1                 0                  0                      0   
3               0                 0                  0                      0   
4               0                 0                  0                      0   

   Item_Type_Soft Drinks  Item_Type_Starchy Foods  
0                      0                        0  
1                      1                        0  
2                      0                        0  
3                      0                        0  
4                      0                        0  

[5 rows x 29 columns]

In [84]:
from sklearn.model_selection import train_test_split

In [85]:
X_train, X_test, y_train, y_test = train_test_split(data_Train_X, data_Train_Y, random_state=1, test_size=0.3)

In [86]:
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(5966, 29)
(5966,)
(2557, 29)
(2557,)


### 8. Linear Regression Model

In [87]:
from sklearn import model_selection, metrics
from sklearn.linear_model import LinearRegression

lin_model = LinearRegression()
lin_model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [88]:
coef1 = pd.DataFrame(lin_model.coef_, index=X_train.columns)

In [89]:
coef1

0
Item_Weight                         2.346934
Item_Visibility                  -427.172241
Item_MRP                           15.661903
Item_Fat_Content_Non-Edible        25.270241
Item_Fat_Content_Regular           29.964324
Outlet_Location_Type_Tier 2       134.567688
Outlet_Location_Type_Tier 3       167.013109
Outlet_Size_1                      75.783233
Outlet_Size_2                    -151.762997
Item_Category_FD                   95.735651
Item_Category_NC                   25.270241
Outlet_Type_Supermarket Type1    1913.455769
Outlet_Type_Supermarket Type2    1403.702519
Outlet_Type_Supermarket Type3    3126.338381
Item_Type_Breads                   49.163741
Item_Type_Breakfast                26.004687
Item_Type_Canned                   19.799580
Item_Type_Dairy                  -103.677153
Item_Type_Frozen Foods            -17.974059
Item_Type_Fruits and Vegetables    34.064720
Item_Type_Hard Drinks              73.880136
Item_Type_Health and Hygiene       50.600310
Item_Type_Household                 9.705737
Item_Type_Meat                      9.769784
Item_Type_Others                  -35.035806
Item_Type_Seafood                 339.654008
Item_Type_Snack Foods              30.698454
Item_Type_Soft Drinks              -7.852101
Item_Type_Starchy Foods            35.609531

#### Intuition from the Co-efficient values:

After looking at the values of Co-efficients, we can see that
1. Each item Sales largely depends upon the Store in which we are selling it. 
2. Location of the outlet influences most of the Item Sales. 

We should clearly keep in mind that all these variables are inter-connected. We can't really identify how much an individual variable effects the Item_Outlet_Sales. Instead we assume that by keeping all the remaining variables fixed, how much a unit change in one variable effects overall Item_Outlet_Sales. 

In [90]:
data_train_predictions = lin_model.predict(X_test)

In [91]:
# Lets have a look at some of the predictions

data_train_predictions[:8]

array([ 626.40908395, 1206.70339064, 3198.06673974, 1461.2590831 ,
        799.0132219 , 3316.22657296,  998.98423998, 2360.00196238])

In [92]:
# We will evaluate the model performance with the RMSE value for the  model.
# Let's find the RMSE for linear regression model we built.

print("RMSE:", np.sqrt(metrics.mean_squared_error(y_test, data_train_predictions)))

RMSE: 1124.564552427027


In [93]:
lin_model.score(X_test,y_test)

0.5652494395089176

In [94]:
print(metrics.r2_score(y_test,data_train_predictions))

0.5652494395089176


### 9.  XGBoost Regressor

In [95]:
from xgboost.sklearn import XGBRegressor

In [96]:
xgb = XGBRegressor(n_estimators=100,
                    learning_rate=0.08,
                    gamma=0,
                    subsample=0.75,
                    colsample_bytree=1,
                    max_depth=7)
xgb.fit(X_train.values, y_train.values)

[23:14:54] WARNING: src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bynode=1, colsample_bytree=1, gamma=0,
       importance_type='gain', learning_rate=0.08, max_delta_step=0,
       max_depth=7, min_child_weight=1, missing=None, n_estimators=100,
       n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=None, subsample=0.75, verbosity=1)

In [97]:
y_pred = xgb.predict(X_test.values)

In [98]:
print('RMSE:',np.sqrt(metrics.mean_squared_error(y_test,y_pred)))   

RMSE: 1114.6019482875968


In [99]:
print(metrics.r2_score(y_test,y_pred))   

0.57291829639282


**Observation** : The RMSE of the model has reduced slightly from 1124 to 1114.  <br>
Lets use Grid Search to find the best hyperparameters for the XGBoost model.

### 9.1  XGBoost Regressor

In [100]:
from sklearn.model_selection import GridSearchCV

In [101]:
# param_grid is a dictionary containing different parameters for XGBoost

param_grid = {"max_depth": [3,5],
              "n_estimators": [50,100,200] , 
              "gamma": [0.1, 0.2], 
              "learning_rate": [0.03,0.05],
              "min_child_weight": [1,2], 
              "colsample_bytree": [1], 
              "subsample": [0.8, 0.85]}

# Performing Grid Search on xgb regressor and param_grid parameters

grid_search = GridSearchCV(xgb,
                          param_grid=param_grid,
                          cv=3,
                          n_jobs=-1,
                          scoring='neg_mean_squared_error',
                          verbose=2)

In [102]:
grid_search.fit(X_train.values,y_train.values)

Fitting 3 folds for each of 96 candidates, totalling 288 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  33 tasks      | elapsed:   34.3s
[Parallel(n_jobs=-1)]: Done 154 tasks      | elapsed:  1.2min
[Parallel(n_jobs=-1)]: Done 288 out of 288 | elapsed:  1.9min finished


[23:16:48] WARNING: src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


GridSearchCV(cv=3, error_score='raise-deprecating',
       estimator=XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bynode=1, colsample_bytree=1, gamma=0,
       importance_type='gain', learning_rate=0.08, max_delta_step=0,
       max_depth=7, min_child_weight=1, missing=None, n_estimators=100,
       n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=None, subsample=0.75, verbosity=1),
       fit_params=None, iid='warn', n_jobs=-1,
       param_grid={'max_depth': [3, 5], 'n_estimators': [50, 100, 200], 'gamma': [0.1, 0.2], 'learning_rate': [0.03, 0.05], 'min_child_weight': [1, 2], 'colsample_bytree': [1], 'subsample': [0.8, 0.85]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='neg_mean_squared_error', verbose=2)

In [103]:
print(grid_search.best_params_)

{'colsample_bytree': 1, 'gamma': 0.1, 'learning_rate': 0.03, 'max_depth': 3, 'min_child_weight': 1, 'n_estimators': 200, 'subsample': 0.85}


In [104]:
y_pred = grid_search.predict(X_test.values)

In [105]:
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test,y_pred)))    # Measuring the error

RMSE: 1071.8866724238333


In [106]:
print(metrics.r2_score(y_test,y_pred))                  # Checking the R2 score.

0.6050254501370484


**Observation**: The RMSE of the model is 1072 after tuning the XGBoost. 

### 10. Model Evaluation 

In this phase of the project we're going to evaluate the models that we've built so far.
We will compare the RMSE obtained from different models.

RMSE for Linear Regression - 1124   <br>
RMSE for XGBoost - 1115   <br>
RMSE for XGBoost with GridSearchCV - 1072


XGBoost with GridSearchCV performed the best and we will use it to make predictions from the test data.

Let's glue these Item_Outlet_Sales values to our test dataset and create a csv file.

**Before make predictions on the Test Data, we need to preprocess the data in the same steps as the Train data**

In [107]:
print(data_Test.columns)                        
data_Test.shape

Index(['Item_Weight', 'Item_Fat_Content', 'Item_Visibility', 'Item_Type',
       'Item_MRP', 'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type',
       'Item_Category'],
      dtype='object')


(5681, 9)

In [108]:
data_Train.drop(['Item_Outlet_Sales'], axis=1, inplace=True)

In [109]:
# Join train and test

data = pd.concat([data_Train, data_Test])

In [110]:
categorical_df = data[cat_feat_var]

In [111]:
# Lets convert 'Outlet_Size' to string data type

categorical_df['Outlet_Size']=categorical_df['Outlet_Size'].astype(str)

C:\Users\vishw\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [112]:
# Get dummy values for all categorical data

dummy_encoded_data = pd.get_dummies(categorical_df, drop_first = True)
dummy_encoded_data.shape

(14204, 26)

In [113]:
# Merge dummy encoded data with original data

one_hot_encoded_data = pd.concat([dummy_encoded_data, data],1)
one_hot_encoded_data.shape

(14204, 35)

In [114]:
# Remove all categorical data as we now have them dummified

one_hot_encoded_data.drop(categorical_df,1,inplace=True)
one_hot_encoded_data.shape

(14204, 29)

In [115]:
# Extract the test data

one_hot_encoded_Test_data = one_hot_encoded_data[-5681:]

In [116]:
# Save it to a numpy array before model prediction

X_test = one_hot_encoded_Test_data.values

In [117]:
X_test.shape

(5681, 29)

In [118]:
X_train.shape

(5966, 29)

In [119]:
# Make predictions on the test data using XGBoost with GridSearchCV.

y_test_predict=grid_search.predict(X_test)   

In [120]:
y_test_predict.shape

(5681,)

In [121]:
data_Test["Item_Outlet_Sales"] = y_test_predict

In [122]:
data_Test.to_csv("Retail_Output.csv", index_label=False, index=False)